piv/src/m1rec2piv/m1rec2piv.1
.\""
.Dd September 20, 2006
.Dt M1REC2PIV 1  
.Sh NAME
.Nm m1rec2piv
.Nd Create a Personal Identity Verification record, from
a set of finger minutiae records or a set of finger image records.
The PIV record
will meet the criteria specified in NIST Special Publication 800-76-1
in all aspects excepting the signature block length, which will be set to 0.
.Sh SYNOPSIS
.Nm
.Fl h
.Ar hdrfile
.Fl f
.Ar filelist
.Fl o
.Ar datafile
.Op Fl s Ar sigfile
.Fl t
.Ar [FMR | FIR]
.Pp
.Sh DESCRIPTION
The
.Nm
command is used to create a new Personal Identity Verification (PIV) record
from a set of finger minutiae records (FMR) or finger image records (FIR),
and PIV header information.
If the type is 'FMR', the finger view records from all the FMRs are combined
into a single FMR, which is then added to the PIV record after the CBEFF header.
Similarly, if the type is 'FIR', then the finger image view records from all
of the FIRs are combined into a single FIR, which is then added to the PIV
record.
The PIV record complies with NIST Special Publication 800-76-1, and is stored
in the output file. 
The FMRs are expected to comply with the ANSI/INCITS 378-2004 specification.
The FIRs are expected to comply with the ANSI/INCITS 381-2004 specification.
Note that it is possible to create a PIV record that is not compliant, by
either having incorrect information in the header, by having more than
two finger views, or with invalid finger minutiae records.
.Pp
Options:
.Bl -tag
.It Fl h
specifies the file containing the PIV CBEFF header information. This file
is a text file, and is in this form:
.Bl -tag -width "Biometric Creation Date" -compact
.It Cm Patron Header Version
Hex; PIV requires 0x03.
.It Cm SBH Security Options
Hex; PIV requires 0x0D for minutiae records.
.It Cm BDB Length
Decimal; not currently used as the length is calculated.
.It Cm SB Length
Decimal; note that a signature block is not processed by
.Nm .
.It Cm BDB Format Owner
Hex; PIV requires 0x001B.
.It Cm BDB Format Type
Hex; PIV requires 0x0201 for minutiae, 0x0401 for finger image.
.It Cm Biometric Creation Date
Hex; Up to 16 hex digits (8 bytes).
.It Cm Validity Period
Hex; Up to 32 hex digits (16 bytes).
.It Cm Biometric Type
Hex; PIV requires 0x000008 for finger images and minutiae.
.It Cm Biometric Data Type
Hex; PIV requires b100xxxxx for minutiae, b001xxxxx for images.
.It Cm Biometric Data Quality
Decimal.
.It Cm Creator
String; Up to 17 characters in length.
.It Cm FASC-N
Hex; Up to 50 hex digits (25 bytes).
.It Cm Reserved
Hex; PIV requires 0x00000000.
.El
.It Fl f
specifies the file containing a list of files that contain complete FMRs or
FIRs.
.It Fl o
specifies the name of the output file. This file must not exist.
.It Fl s
specifies the name of the file containing the signature block to attach to the
 end of the record. For a properly formed PIV record, the size of this file
should match the length specified in the CBEFF header, although no check is
made by this program.
.It Fl t
specifies the type of biometric data, either FMR or FIR.
.El
.Pp
The
.Nm
program exits with one of the following values:
.Bl -tag -width Ds
.It 0
File was successfully created.
.It -1
An error occurred.
.El
.Sh EXAMPLES
.Bd -literal
m1rec2piv -h hdr.txt -f fmrs.txt -o piv.raw -s sb.raw -t fmr
.Ed
.Sh FILES
Example PIV CBEFF header info file:

0x03 0x0D 0 0 0x001B 0x0201 0x14050C0F11231D5A
0x14050C0F11231D5A14050C0F11231D5A 0x000008 0x80 80 NIST-IAD
0x00000000000000000000000000000000000000000000000000 0x00000000

(The above example is shown as three lines; in the actual file the information
must be on a single line.)
.Sh SEE ALSO
.Xr pivv 1 ,
.Xr prfmr 1 .
.Sh STANDARDS
``Biometric Specification for Personal Identity Verification'', NIST
Special Publication 800-76-1, September 05, 2006.
.Pp
``Finger Minutiae Format for Data Interchange'', ANSI INCITS 378-2004.
``Finger Image-Based Data Interchange Format'', ANSI INCITS 385-2004.
.Sh HISTORY
Created August 30th, 2006 by NIST.

Signature block option added December 9th, 2008 by NIST.


piv/src/m1rec2piv/m1rec2piv.c
/*
* This software was developed at the National Institute of Standards and
* Technology (NIST) by employees of the Federal Government in the course
* of their official duties. Pursuant to title 17 Section 105 of the
* United States Code, this software is not subject to copyright protection
* and is in the public domain. NIST assumes no responsibility  whatsoever for
* its use by other parties, and makes no guarantees, expressed or implied,
* about its quality, reliability, or any other characteristic.
*/
/******************************************************************************/
/* This program will combine several Finger Minutiae Records or a set of
/* Finger Image Records into a PIV compliant record, stored in a file.        */
/*                                                                            */
/* XXX It would be desirable to create the signature block.                   */
/******************************************************************************/

/* Needed by the GNU C libraries for Posix and other extensions */
#define _XOPEN_SOURCE	1

#include <sys/param.h>
#include <sys/queue.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>

#include <biomdimacro.h>
#include <fir.h>
#include <fmr.h>
#include <piv.h>

struct fmrelem {
	struct finger_minutiae_record		*fmr;
	TAILQ_ENTRY(fmrelem)			list;
};

static void
usage(char *name)
{
	printf("usage: %s -h <hdrfile> -f <listfile> -o <outfile>"
	    " [-s <sigfile>] -t [FMR | FIR]\n", name);
	exit (EXIT_FAILURE);
}

/*
 * Convert a string starting with '0x' or '0X' to an array of unsigned ints.
 */
static int
atox(char *s, uint8_t b[], int len)
{
	int i;
	char buf[5];

	if (s[0] != '0' || (toupper(s[1]) != 'X'))
		return (-1);
	for (i = 0; i < len; i++) {
		sprintf(buf, "0x%c%c", s[(i + 1)*2], s[((i + 1)*2) + 1]);
		sscanf(buf, "%hhx", &b[i]);
	}
	
}

/*
 * Read the PIV CBEFF header.
 */
static int
readin_pcr(FILE *infp, struct piv_cbeff_record *pcr)
{
	int count, rval, i;
	char buf[255];

	bzero((void *)pcr, sizeof(struct piv_cbeff_record));
	rval = fscanf(infp, "%hhx %hhx %u %hu %hx %hx",
	    &pcr->patron_header_version,
	    &pcr->sbh_security_options,
	    &pcr->bdb_length,
	    &pcr->sb_length,
	    &pcr->bdb_format_owner,
 	    &pcr->bdb_format_type);
	count = rval;
	if (count != 6)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	   
	rval = fscanf(infp, "%s", buf);
	if (rval != 1)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	if (atox(buf, pcr->biometric_creation_date,
	    BIOMETRIC_CREATION_DATE_LEN) < 0)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	count += rval;

	rval = fscanf(infp, "%s", buf);
	if (rval != 1)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	if (atox(buf, pcr->validity_period, VALIDITY_PERIOD_LEN) < 0)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	count += rval;

	rval = fscanf(infp, "%x %hhx %hhd %17s",
	    &pcr->biometric_type,
	    &pcr->biometric_data_type,
	    &pcr->biometric_data_quality,
	    pcr->creator);
	if (rval != 4)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	count += rval;

	rval = fscanf(infp, "%s", buf);
	if (rval != 1)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	if (atox(buf, pcr->fascn, FASCN_LEN) < 0)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);
	count += rval;

	rval = fscanf(infp, "%x", pcr->reserved);
	if (rval != 1)
		READ_ERR_RETURN("PCR header (read %d of 14 items)", count);

	return (READ_OK);
}

static int
readin_fmrs(FILE *infp, FILE *outfp, struct finger_minutiae_record **fmrp)
{
	FILE *fmrfp = NULL;
	struct fmrelem *fmrelem;
	TAILQ_HEAD(, fmrelem) fmrlist;
	char fmr_filename[MAXPATHLEN + 1];
	struct finger_minutiae_record *fmr;
	struct finger_view_minutiae_record *fvmr;
	int ret;

	TAILQ_INIT(&fmrlist);
	// Read in all the finger minutiae records from their respective
	// files, and place them in a list.
	for (;;) {
		ret = fscanf(infp, "%s", fmr_filename);
		if (ret == EOF)
			break;
		if (ret != 1)
			READ_ERR_OUT(fmr_filename);
		if (new_fmr(FMR_STD_ANSI, &fmr) < 0)
			ALLOC_ERR_OUT("finger minutiae record");

		fmrfp = fopen(fmr_filename, "rb");
		if (fmrfp == NULL)
			ERR_OUT("Could not open %s", fmr_filename);
		if (read_fmr(fmrfp, fmr) != READ_OK)
			READ_ERR_OUT(fmr_filename);
		fclose(fmrfp);

		fmrelem = (struct fmrelem *)malloc(sizeof(struct fmrelem));
		if (fmrelem == NULL)
			ALLOC_ERR_OUT("finger minutiae record list");
		fmrelem->fmr = fmr;
		TAILQ_INSERT_TAIL(&fmrlist, fmrelem, list);
	}
	// Use the first FMR as the master, and add all the finger view
	// minutiae records to it.
	fmrelem = TAILQ_FIRST(&fmrlist);
	TAILQ_REMOVE(&fmrlist, fmrelem, list);
	fmr = fmrelem->fmr;
	free(fmrelem);
	while (!TAILQ_EMPTY(&fmrlist)) {
		fmrelem = TAILQ_FIRST(&fmrlist);
		TAILQ_REMOVE(&fmrlist, fmrelem, list);
		// Fix up the fmr record length
		fmr->record_length += fmrelem->fmr->record_length -
			fmrelem->fmr->record_length_type;

		// The current 378 spec says that the image size in the
		// header is to be the largest of the views. For this
		// program, we want to find the largest given in the headers.
		if (fmrelem->fmr->x_image_size > fmr->x_image_size)
			fmr->x_image_size = fmrelem->fmr->x_image_size;
		if (fmrelem->fmr->y_image_size > fmr->y_image_size)
			fmr->y_image_size = fmrelem->fmr->y_image_size;

		while (!TAILQ_EMPTY(&fmrelem->fmr->finger_views)) {
			fvmr = TAILQ_FIRST(&fmrelem->fmr->finger_views);
			TAILQ_REMOVE(&fmrelem->fmr->finger_views, fvmr, list);
			add_fvmr_to_fmr(fvmr, fmr);
			fmr->num_views++;
		}
		free_fmr(fmrelem->fmr);
	}
	*fmrp = fmr;
	return (READ_OK);

err_out:
	free_fmr(fmr);
	if (fmrfp != NULL)
		fclose(fmrfp);
	return (READ_ERROR);
}

static int
readin_firs(FILE *infp, FILE *outfp, struct finger_image_record **firp)
{
	FILE *firfp = NULL;
	char fir_filename[MAXPATHLEN + 1];
	struct finger_image_record *headfir = NULL;
	struct finger_image_record *fir = NULL;
	struct finger_image_view_record **fivrs;
	struct finger_image_view_record *fivr;
	int ret;
	int rcount;
	int i;

	// Read in the first finger image record 
	ret = fscanf(infp, "%s", fir_filename);
	if (ret != 1)
		READ_ERR_OUT("FIR list file");
	if (new_fir(FMR_STD_ANSI, &headfir) < 0)
		ALLOC_ERR_OUT("finger image record");
	firfp = fopen(fir_filename, "rb");
	if (firfp == NULL)
		ERR_OUT("Could not open %s", fir_filename);
	if (read_fir(firfp, headfir) != READ_OK)
		READ_ERR_OUT(fir_filename);
	fclose(firfp);

	// Read in the remaining FIRs and add their view records to the 
	// first FIR
	for (;;) {
		ret = fscanf(infp, "%s", fir_filename);
		if (ret == EOF)
			break;
		if (ret != 1)
			READ_ERR_OUT(fir_filename);
		if (new_fir(FMR_STD_ANSI, &fir) < 0)
			ALLOC_ERR_OUT("finger image record");

		firfp = fopen(fir_filename, "rb");
		if (firfp == NULL)
			ERR_OUT("Could not open %s", fir_filename);
		if (read_fir(firfp, fir) != READ_OK)
			READ_ERR_OUT(fir_filename);
		fclose(firfp);

		// Pull off the FIVRs
		rcount = get_fivr_count(fir);
		fivrs = (struct finger_image_view_record **) malloc(
		    rcount * sizeof(struct finger_image_view_record *));
		if (fivrs == NULL)
			ALLOC_ERR_EXIT("FIVR Array");
		if (get_fivrs(fir, fivrs) != rcount)
			ERR_OUT("getting FIVRs from FIR");

		for (i = 0; i < rcount; i++) {
			if (new_fivr(&fivr) < 0)
				ALLOC_ERR_OUT("finger image view record copy");
			copy_fivr(fivrs[i], fivr);
			add_fivr_to_fir(fivr, headfir);
		}
		free_fir(fir);
		fir = NULL;
	}

	*firp = headfir;
	return (READ_OK);

err_out:
	free_fir(headfir);
	if (fir != NULL)
		free_fir(fir);
	if (firfp != NULL)
		fclose(firfp);
	return (READ_ERROR);
}

int
main(int argc, char *argv[])
{
	struct piv_cbeff_record pcr;
	struct finger_minutiae_record *fmr = NULL;
	struct finger_image_record *fir = NULL;
	int ch;
	int exit_code = EXIT_FAILURE;
	int hflag, fflag, oflag, sflag, tflag;
	FILE *h_fp, *f_fp, *o_fp, *s_fp;
	struct stat sb;
	char f_opt_filename[MAXPATHLEN + 1];
	int type;
	int s_size;
	char *s_buf;

	hflag = oflag = fflag = tflag = sflag = 0;
	s_buf = NULL;
	while ((ch = getopt(argc, argv, "h:f:o:s:t:")) != -1) {
		switch (ch) {
			case 'h' :
				if ((h_fp = fopen(optarg, "r")) == NULL)
					OPEN_ERR_EXIT(optarg);
				hflag = 1;
				break;
			case 'f' :
				if ((f_fp = fopen(optarg, "r")) == NULL)
					OPEN_ERR_EXIT(optarg);
				strncpy(f_opt_filename, optarg, MAXPATHLEN);
				fflag = 1;
				break;
			case 'o' :
				if (stat(optarg, &sb) == 0)
					ERR_EXIT("File '%s' exists, remove it first", optarg);
				if ((o_fp = fopen(optarg, "wb")) == NULL)
					OPEN_ERR_EXIT(optarg);
				oflag = 1;
				break;
			case 's' :
				if (stat(optarg, &sb) != 0)
					ERR_EXIT("Could not find %s", optarg);
				if ((s_fp = fopen(optarg, "r")) == NULL)
					OPEN_ERR_EXIT(optarg);
				s_size = sb.st_size;
				s_buf = (char *)malloc(s_size);
				if (s_buf == NULL)
					ALLOC_ERR_EXIT("Signature buffer");
				if (fread(s_buf, 1, s_size, s_fp) != s_size)
					ERR_EXIT("Could not read %s", optarg);
				fclose(s_fp);
				sflag = 1;
				break;
			case 't' :
				if (strncasecmp(optarg, FMR_FORMAT_ID,
				    strlen(FMR_FORMAT_ID)) == 0)
					type = PIVFMR;
				else if (strncasecmp(optarg, FIR_FORMAT_ID,
				    strlen(FIR_FORMAT_ID)) == 0)
					type = PIVFIR;
				else
					usage(argv[0]);
				tflag = 1;
				break;
			default :
				usage(argv[0]);
				break;	// not reached
		}
	}

	if (!hflag || !fflag || !oflag || !tflag)
		usage(argv[0]);

	// Create the PIV CBEFF header
	if (readin_pcr(h_fp, &pcr) != READ_OK)
		READ_ERR_OUT("PIV CBEFF header");

	switch (type) {
		case PIVFMR:
			if (readin_fmrs(f_fp, o_fp, &fmr) != READ_OK)
				READ_ERR_OUT("FMRs");
			pcr.bdb_length = fmr->record_length;
			break;
		case PIVFIR:
			if (readin_firs(f_fp, o_fp, &fir) != READ_OK)
				READ_ERR_OUT("FIRs");
			pcr.bdb_length = fir->record_length;
			break;
	}
		
	if (piv_write_pcr(o_fp, &pcr) != WRITE_OK)
		WRITE_ERR_OUT("CBEFF Header into PIV file");

	switch (type) {
		case PIVFMR:
			if (write_fmr(o_fp, fmr) != WRITE_OK)
				WRITE_ERR_OUT("FMR into PIV file");
			break;
		case PIVFIR:
			if (write_fir(o_fp, fir) != WRITE_OK)
				WRITE_ERR_OUT("FIR into PIV file");
			break;
	}

	/* Write out the signature, if specified */
	if (sflag)
		if (fwrite(s_buf, 1, s_size, o_fp) != s_size)
			WRITE_ERR_OUT("Signature block");

	exit_code = EXIT_SUCCESS;

err_out:
	if (fmr != NULL)
		free_fmr(fmr);
	if (fir != NULL)
		free_fir(fir);
	if (s_buf != NULL)
		free(s_buf);
	fclose(h_fp);
	fclose(f_fp);
	fclose(o_fp);
	
	exit(exit_code);
}
